File: src/main/java/com/hazza/LeetCodeByJava/SegmentOne/No4MedianOfTwoSortedArrays.java
package com.hazza.LeetCodeByJava.SegmentOne;

/**
 * https://leetcode.com/problems/median-of-two-sorted-arrays/#/description
 */
public class No4MedianOfTwoSortedArrays {
    public double findMedianSortedArrays(int[] A, int[] B) {
        int m = A.length;
        int n = B.length;
        //若m+n为奇数，则l=r
        //若m+n为偶数，则l为偏左的中位数，r为偏右的中位数
        int l = (m + n + 1) / 2;
        int r = (m + n + 2) / 2;
        return (findKth(A, 0, B, 0, l) + findKth(A, 0, B, 0, r)) / 2.0;
    }

    private double findKth(int[] A, int aStart, int[] B, int bStart, int k) {
        if (aStart > A.length - 1) return B[bStart + k - 1];	//A为空或A的最大元素小于B[bStart]
        if (bStart > B.length - 1) return A[aStart + k - 1];	//B为空或B的最大元素小于A[aStart]
        if (k == 1) return Math.min(A[aStart], B[bStart]);	//k=1,只剩一个数字，取小的

        //防止A或B数组过短
        int aMid = Integer.MAX_VALUE, bMid = Integer.MAX_VALUE;
        if (aStart + k/2 - 1 < A.length) aMid = A[aStart + k/2 - 1];
        if (bStart + k/2 - 1 < B.length) bMid = B[bStart + k/2 - 1];

        if (aMid < bMid) return findKth(A, aStart + k/2, B, bStart, k - k/2);
        else return findKth(A, aStart, B, bStart + k/2, k - k/2);
    }
}

File: src/main/python/com/hazza/leetcode_by_python/SegmentOne/No4_Median_of_Two_Sorted_Arrays.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/4 21:32
# https://leetcode.com/problems/median-of-two-sorted-arrays/
import sys
from typing import List


class Solution:
    def findKth(self, nums1, m, start1, nums2, n, start2, k):
        if start1 >= m:
            return nums2[start2 + k -1]
        if start2 >= n:
            return nums1[start1 + k - 1]
        if k == 1:
            return min(nums1[start1], nums2[start2])

        mid1 = sys.maxsize
        mid2 = sys.maxsize
        half = int(k / 2)
        if start1 + half - 1 < m:
            mid1 = nums1[start1 + half - 1]
        if start2 + half - 1 < n:
            mid2 = nums2[start2 + half - 1]

        if mid1 < mid2:
            return self.findKth(nums1, m, start1 + half, nums2, n, start2, k - half)
        else:
            return self.findKth(nums1, m, start1, nums2, n, start2 + half, k - half)

    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m = len(nums1)
        n = len(nums2)
        l = int((m + n + 1) / 2)
        r = int((m + n + 2) / 2)

        return (self.findKth(nums1, m, 0, nums2, n, 0, l) + self.findKth(nums1, m, 0, nums2, n, 0, r)) / 2.0


if __name__ == '__main__':
    solution = Solution()
    assert solution.findMedianSortedArrays([1, 3], [2]) == 2.0
    assert solution.findMedianSortedArrays([1, 2], [3, 4]) == 2.5
